plugins/analysis-icu/src/main/java/org/elasticsearch/index/analysis/IcuTokenizerFactory.java

```diff
@@ -19,28 +19,119 @@
+        }
+    }
+
+    public ICUTokenizerConfig getIcuConfig(Environment env, Settings settings) {
+        Map<Integer, String> tailored = new HashMap<>();
+
+        try {
+            String[] ruleFiles = settings.getAsArray(RULE_FILES);
+            for (String scriptAndResourcePath : ruleFiles) {
+                int colonPos = scriptAndResourcePath.indexOf(":");
+                if (colonPos == -1 || colonPos == scriptAndResourcePath.length() - 1) {
+                    throw new IllegalArgumentException(RULE_FILES + " should contain comma-separated \"code:rulefile\" pairs");
+                }
+
+                String scriptCode = scriptAndResourcePath.substring(0, colonPos).trim();
+                String resourcePath = scriptAndResourcePath.substring(colonPos + 1).trim();
+                tailored.put(UCharacter.getPropertyValueEnum(UProperty.SCRIPT, scriptCode), resourcePath);
+            }
+
+            if (tailored.isEmpty()) {
+                return null;
+            } else {
+                final BreakIterator breakers[] = new BreakIterator[UScript.CODE_LIMIT];
+                for (Map.Entry<Integer, String> entry : tailored.entrySet()) {
+                    int code = entry.getKey();
+                    String resourcePath = entry.getValue();
+                    breakers[code] = parseRules(resourcePath, env);
+                }
+                // cjkAsWords is not configurable yet.
+                ICUTokenizerConfig config = new DefaultICUTokenizerConfig(true) {
+                    @Override
+                    public BreakIterator getBreakIterator(int script) {
+                        if (breakers[script] != null) {
+                            return (BreakIterator) breakers[script].clone();
+                        } else {
+                            return super.getBreakIterator(script);
+                        }
+                    }
+                };
+                return config;
+            }
+        } catch (Throwable t) {
+            throw new ElasticsearchException("failed to load ICU rule files", t);
+        }
     }
 
+    //parse a single RBBi rule file
+    private BreakIterator parseRules(String filename, Environment env) throws IOException {
+        StringBuilder rules = new StringBuilder();
+
+        final Path path = env.configFile().resolve(filename);
+        BufferedReader reader = new BufferedReader(
+                FileSystemUtils.newBufferedReader(path.toUri().toURL(), StandardCharsets.UTF_8));
+
+        try {
+            String line = null;
+            while ((line = reader.readLine()) != null) {
+                if (!line.startsWith("#")) {
+                    rules.append(line);
+                }
+                rules.append('\n');
+            }
+
+        } catch (Throwable t) {
+            throw new IllegalArgumentException("couldn't parse the rule file : [" + filename + "]");
+        } finally {
+            reader.close();
```

Comment:
We can rely on auto-closeables here (i.e. we could use just the try-with-resource statement).